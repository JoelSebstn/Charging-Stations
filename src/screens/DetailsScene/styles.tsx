import { StyleSheet } from "react-native";

export const Styles = StyleSheet.create({
    alignCentre: { flex: 1, alignItems: 'center' },
    textFont22: { fontSize: 22 },
    detailsSceneView: { flexDirection: 'column', justifyContent: 'center', marginTop: 40 },
    nameStyle: { fontSize: 28, textAlign: 'center', fontWeight: '700' },
    addressStyle: { fontSize: 24, textAlign: 'center', },
    margin20: { marginBottom: 20 },
    margin40: { marginBottom: 40 },
    flexRow: { flexDirection: 'row' },
    horizontalLine: {
        borderBottomColor: 'black',
        borderBottomWidth: 2,
    }

});