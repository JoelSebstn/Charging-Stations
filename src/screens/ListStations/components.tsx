import { ActivityIndicator, Text, TouchableOpacity, View } from "react-native";
import { Styles } from "./style";
import { useNavigation } from "@react-navigation/native";

export function HeaderRow() {
    return (<><View style={Styles.flexRow}>
        <View style={Styles.alignCentre}><Text style={Styles.textFont22}>Name</Text></View>
        <View style={Styles.alignCentre}><Text style={Styles.textFont22}> Status</Text></View>
    </View>
        <View
            style={Styles.horizontalLine}
        /></>
    )
}
export function RenderItem(item: any) {
    const data = item["item"].data();
    const stationId = item["item"].data().id;
    const navigation: any = useNavigation();
    return (
        <TouchableOpacity onPress={() => navigation.navigate('DetailsSCene', { stationId: stationId })}>
            <View style={{ flexDirection: 'row', marginVertical: 12 }}>
                <View style={Styles.alignCentre}><Text style={Styles.textFont22}>{data.name}</Text></View>
                <View style={Styles.alignCentre}><Text style={Styles.textFont22}> {data.status}</Text></View>
            </View>
        </TouchableOpacity>
    )

}
export function renderFooter(isLoading: any) {
    try {
        // Check If Loading
        if (isLoading) {
            return (
                <ActivityIndicator />
            )
        }
        else {
            return null;
        }
    }
    catch (error) {
        console.log(error);
    }
};