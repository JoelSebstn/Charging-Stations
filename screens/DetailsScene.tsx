import { Text, View } from "react-native";
import { Styles } from "../styles/ListScreenStyles";

export function DetailsScene({ route }: any) {
    const station = route.params["station"];
    return (
        <View style={Styles.detailsSceneView}>
            <View >
                <Text style={Styles.nameStyle}>{station["name"]}</Text>
                <Text style={Styles.addressStyle}>{station["address"]}</Text>
            </View>
            <View
                style={Styles.margin40}
            />
            <View style={Styles.flexRow}>
                <View style={Styles.alignCentre}><Text style={Styles.textFont22}>Opening Hours:</Text></View>
                <View style={Styles.alignCentre}><Text style={Styles.textFont22}> {station["opening hours"]}</Text></View>
            </View >
            <View
                style={Styles.margin20}
            />
            <View style={Styles.flexRow}>
                <View style={Styles.alignCentre}><Text style={Styles.textFont22}>Operator:</Text></View>
                <View style={Styles.alignCentre}><Text style={Styles.textFont22}> {station["operator"]}</Text></View>
            </View >
            <View
                style={Styles.margin20}
            />
            <View style={Styles.flexRow}>
                <View style={Styles.alignCentre}><Text style={Styles.textFont22}>Parking Access:</Text></View>
                <View style={Styles.alignCentre}><Text style={Styles.textFont22}>{station["parking access"] == true ? 'Yes' : 'No'}</Text></View>
            </View >
            <View
                style={Styles.margin20}
            />
            <View style={Styles.flexRow}>
                <View style={Styles.alignCentre}><Text style={Styles.textFont22}>Status:</Text></View>
                <View style={Styles.alignCentre}><Text style={Styles.textFont22}> {station["status"]}</Text></View>
            </View >
        </View >)

}