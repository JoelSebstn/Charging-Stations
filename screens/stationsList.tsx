import { useEffect, useState } from "react";
import { FlatList, SafeAreaView, Text, TouchableOpacity, View } from "react-native"
import { tempStationsData } from "../data/stations";
import { Styles } from "../styles/ListScreenStyles";
import firestore from '@react-native-firebase/firestore';



export function ListStations({ navigation }: any) {
    const [stationsData, setStationsData] = useState<any[]>(tempStationsData["stations"])
    const [lastVisible, setLastVisible] = useState(null);
    const [isLoading, setIsLoading] = useState(false);
    const [isRefreshing, setIsRefreshing] = useState(false);
    useEffect(() => {
        fetchStations();
    });
    async function fetchStations() {
        try {
            const locationCollection = await firestore().collection('locations').get();
            console.log('locationCollection: ', locationCollection);

        }
        catch (error) {
            console.log("errorCAtched:", error);
        }
    }
    return (
        <SafeAreaView >
            <View>
                <View style={Styles.flexRow}>
                    <View style={Styles.alignCentre}><Text style={Styles.textFont22}>Name</Text></View>
                    <View style={Styles.alignCentre}><Text style={Styles.textFont22}> Status</Text></View>
                </View>
                <View
                    style={{
                        borderBottomColor: 'black',
                        borderBottomWidth: 2,
                    }}
                />
                <FlatList
                    // Data 
                    data={stationsData}
                    // Render Items
                    renderItem={({ item }) => (
                        <TouchableOpacity onPress={() => navigation.navigate('DetailsSCene', { station: item })}>
                            <View style={{ flexDirection: 'row', marginVertical: 12 }}>
                                <View style={Styles.alignCentre}><Text style={Styles.textFont22}>{item.name}</Text></View>
                                <View style={Styles.alignCentre}><Text style={Styles.textFont22}> {item.status}</Text></View>
                            </View>
                        </TouchableOpacity>
                    )
                    }
                    // Element Key
                    keyExtractor={(item, index) => String(index)}
                    // Header (Title)
                    // ListHeaderComponent={this.renderHeader}
                    // Footer (Activity Indicator)
                    // ListFooterComponent={this.renderFooter}
                    // On End Reached (Takes in a function)
                    // onEndReached={this.retrieveMore}
                    // How Close To The End Of List Until Next Data Request Is Made
                    onEndReachedThreshold={0}
                // Refreshing (Set To True When End Reached)
                // refreshing={this.state.refreshing}
                /></View >
        </SafeAreaView >
    )
}