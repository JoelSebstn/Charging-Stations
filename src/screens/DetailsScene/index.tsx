import { Text, View } from "react-native";
import { useEffect, useState } from "react";
import firestore from '@react-native-firebase/firestore';
import { Styles } from "./styles";
import { DetailsRow } from "./components";

export function DetailsScene({ route }: any) {
    const stationId = route.params["stationId"];
    const [stationDetails, setStationDetails] = useState<any>();
    useEffect(() => {
        fetchStations();
    }, []);
    async function fetchStations() {
        try {
            await firestore().collection('locations').doc(stationId)
                .onSnapshot(documentSnapshot => {
                    setStationDetails(documentSnapshot.data());
                });
        }
        catch (error) {
            console.log("errorCAtched:", error);
        }
    }
    return (
        stationDetails && <View style={Styles.detailsSceneView} >
            <View style={Styles.margin40} >
                <Text style={Styles.nameStyle}>{stationDetails!['name'] ?? ''}</Text>
                <Text style={Styles.addressStyle}>{stationDetails["addressDetail"]["address"]},{stationDetails["addressDetail"]["city"]}</Text>
            </View>
            <DetailsRow rowName={'Opening Hours'} data={stationDetails["openingHours"]} />
            <DetailsRow rowName={'Operator'} data={stationDetails["operator"]} />
            <DetailsRow rowName={'Parking Access'} data={stationDetails["parkingAccess"]} />
            <DetailsRow rowName={'Status'} data={stationDetails["status"]} />
        </View >)

}