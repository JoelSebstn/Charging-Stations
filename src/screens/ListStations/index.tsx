import { useEffect, useState } from "react";
import { FlatList, SafeAreaView, View } from "react-native"
import { HeaderRow } from "../../components/headerRow/headerRow";
import { RenderItem } from "../../components/renderItem";
import { renderFooter } from "../../components/renderFooter";
import { anonymousAuthentication, fetchMoreStationsFromFirebase, fetchStationsFromFirebase } from "../../utilities/firebaseServices";

export function ListStations({ }, { navigation }: any) {
    const [stationsData, setStationsData] = useState<any[]>()
    const [lastVisible, setLastVisible] = useState(null);
    const [isLoading, setIsLoading] = useState(false);
    useEffect(
        () => {
            fetchStations();
        },
        []
    );
    async function fetchStations() {
        try {
            await anonymousAuthentication();
            fetchStationsFromFirebase(setStationsData, setLastVisible)
        }
        catch (error) {
            console.log("error:", error);
        }
    }
    async function fetchMore() {
        try {
            fetchMoreStationsFromFirebase(setIsLoading, setStationsData, setLastVisible, lastVisible, stationsData)
        } catch (error) {
            console.log("error:", error);
        }
    }
    return (
        <SafeAreaView >
            <HeaderRow />
            <FlatList
                data={stationsData}
                renderItem={({ item }) => (
                    <RenderItem item={item.data()} navigation={navigation} />)
                }
                keyExtractor={(item, index) => String(index)}
                ListFooterComponent={renderFooter(isLoading)}
                onEndReached={fetchMore}
                onEndReachedThreshold={3}
            />
        </SafeAreaView >
    )
}