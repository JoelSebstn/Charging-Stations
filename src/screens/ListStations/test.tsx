

import renderer from 'react-test-renderer';
import { RenderItem } from './components/rederItem';
import { render, fireEvent } from '@testing-library/react-native';

it(`renders button correctly`, () => {
    const data = { "id": 123, "name": "TEst Name", "status": "test status" }
    const tree = renderer.create(<RenderItem item={data} />).toJSON();
    expect(tree).toMatchSnapshot();
})
it('should render the name correctly', () => {

    const data = { "id": 123, "name": "TEst Name", "status": "test status" }
    const { getByText } = render(<RenderItem item={data} />);
    const label = getByText("TEst Name");
    expect(label).toBeTruthy();
});
it('should render the status correctly', () => {

    const data = { "id": 123, "name": "TEst Name", "status": "test status" }
    const { getByText } = render(<RenderItem item={data} />);
    const status = getByText("test status");
    expect(status).toBeTruthy();
});