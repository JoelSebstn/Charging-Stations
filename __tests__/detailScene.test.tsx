import { NavigationContainer } from "@react-navigation/native";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import { ListStations } from "../src/screens/ListStations";
import renderer from 'react-test-renderer';
import React from "react";
import { DetailsScene } from "../src/screens/DetailsScene";

jest.mock('../src/utilities/firebaseServices.tsx', () => ({
    fetchStationDetails: jest.fn(() => Promise.resolve('Mocked Data')),
}));
const mockSnapshotData =
{
    id: 1,
    "name": 'Job Richie',
    "status": "Available",
    "openingHours": "9:00AM",
    "operator": "Jack Mack Shell",
    "parkingAccess": "True",
    "addressDetail": { "address": "Home", "city": "Rome" }
};
const mockRoute = { "params": { "stationId": "123" } };
const setStateMock = jest.fn();
jest.spyOn(React, 'useState').mockReturnValue([mockSnapshotData, setStateMock]);
describe("Test details screen", () => {
    it('screen renders name correctly', async () => {
        const { getByText } = render(
            <NavigationContainer><DetailsScene route={mockRoute} /></NavigationContainer>,);
        expect(getByText('Job Richie')).toBeDefined();
        await waitFor(() => {
        });
    })
    it('screen renders status correctly', () => {
        const { getByText } = render(
            <NavigationContainer><DetailsScene route={mockRoute} /></NavigationContainer>,);
        expect(getByText('Available')).toBeTruthy();
    })
    it('screen renders time correctly', () => {
        const { getByText } = render(
            <NavigationContainer><DetailsScene route={mockRoute} /></NavigationContainer>,);
        expect(getByText('9:00AM')).toBeTruthy();
    })
    it('screen renders operator correctly', () => {
        const { getByText } = render(
            <NavigationContainer><DetailsScene route={mockRoute} /></NavigationContainer>,);
        expect(getByText('Jack Mack Shell')).toBeTruthy();
    })
    it('screen renders address correctly', () => {
        const { getByText } = render(
            <NavigationContainer><DetailsScene route={mockRoute} /></NavigationContainer>,);
        expect(getByText('Home,Rome')).toBeTruthy();
    })
    it("Matches screen Snapshot", () => {
        const domTree = renderer.create(
            <NavigationContainer><DetailsScene route={mockRoute} /></NavigationContainer>).toJSON();
        expect(domTree).toMatchSnapshot();
    });
})