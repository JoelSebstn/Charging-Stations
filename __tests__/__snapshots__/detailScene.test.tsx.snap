// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test details screen Matches screen Snapshot 1`] = `
<View
  style={
    {
      "flexDirection": "column",
      "justifyContent": "center",
      "marginTop": 40,
    }
  }
>
  <View
    style={
      {
        "marginBottom": 40,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 28,
          "fontWeight": "700",
          "textAlign": "center",
        }
      }
    >
      Job Richie
    </Text>
    <Text
      style={
        {
          "fontSize": 24,
          "textAlign": "center",
        }
      }
    >
      Home
      ,
      Rome
    </Text>
  </View>
  <View
    style={
      [
        {
          "flexDirection": "row",
        },
        {
          "marginBottom": 20,
          "marginHorizontal": 8,
        },
      ]
    }
  >
    <View
      style={
        {
          "alignItems": "flex-start",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 22,
          }
        }
      >
        Opening Hours
        :
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "flex-start",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 22,
          }
        }
      >
         
        9:00AM
      </Text>
    </View>
  </View>
  <View
    style={
      [
        {
          "flexDirection": "row",
        },
        {
          "marginBottom": 20,
          "marginHorizontal": 8,
        },
      ]
    }
  >
    <View
      style={
        {
          "alignItems": "flex-start",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 22,
          }
        }
      >
        Operator
        :
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "flex-start",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 22,
          }
        }
      >
         
        Jack Mack Shell
      </Text>
    </View>
  </View>
  <View
    style={
      [
        {
          "flexDirection": "row",
        },
        {
          "marginBottom": 20,
          "marginHorizontal": 8,
        },
      ]
    }
  >
    <View
      style={
        {
          "alignItems": "flex-start",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 22,
          }
        }
      >
        Parking Access
        :
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "flex-start",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 22,
          }
        }
      >
         
        True
      </Text>
    </View>
  </View>
  <View
    style={
      [
        {
          "flexDirection": "row",
        },
        {
          "marginBottom": 20,
          "marginHorizontal": 8,
        },
      ]
    }
  >
    <View
      style={
        {
          "alignItems": "flex-start",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 22,
          }
        }
      >
        Status
        :
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "flex-start",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 22,
          }
        }
      >
         
        Available
      </Text>
    </View>
  </View>
</View>
`;
